vNext: Add migration step for publication-dissertation (#920)
inveniosoftware/docs-invenio-rdm: docs/releases/v14/upgrade-v14.0.md

See also the [documentation on how to configure the new module](../../operate/customize/code_archival.md).

That's it, you have upgraded to InvenioRDM v14!

## Align "Thesis" and "Dissertation" resource types — optional

> Your upgrade is complete. This last section describes an **entirely optional** change that is not part of the upgrade. Nothing here affects your instance unless you choose to run it, and you can do so at any later time. Because resource types are a highly visible and commonly customized vocabulary, we suggest rather than impose this change to decide together with your instance's stakeholders (librarians, curators) whether it fits your data before applying it.

Reviewer: ```suggestion
Your upgrade is complete. This last section describes an **entirely optional** change that is not part of the upgrade. Nothing here affects your instance unless you choose to run it, and you can do so at any later time. Because resource types are a highly visible and commonly customized vocabulary, we suggest rather than impose this change. Decide together with your instance's stakeholders (librarians, curators) whether it fits your data before applying it.
```

### What changed and why

With InvenioRDM v14, we consolidated the distinction between "Thesis" and "Dissertation" resource types, as these are treated as [the same category](https://datacite-metadata-schema.readthedocs.io/en/4.7/appendices/appendix-1/resourceTypeGeneral/#dissertation) in the DataCite metadata standard. The default InvenioRDM resource types now use "Publication / Thesis" as the primary term in the deposit form, backed by the id `publication-dissertation`, and drop the separate `publication-thesis` entry. When minting DataCite DOIs, InvenioRDM maps this resource type to the standard `Dissertation` value.

![](./imgs/dissertation-thesis.jpg)

The motivation is interoperability. The former `publication-thesis` type minted its DataCite DOIs with a custom `Text` value, whereas a standard DataCite value (`Dissertation`) applies. Staying close to the DataCite schema is a core goal of InvenioRDM, so aligning existing records with `publication-dissertation` keeps your metadata interoperable. See DataCite's [`Dissertation` definition](https://datacite-metadata-schema.readthedocs.io/en/4.7/appendices/appendix-1/resourceTypeGeneral/#dissertation) for its intended scope.

> That said, your data is yours. If you deliberately keep both types (for example `publication-dissertation` for PhD work and `publication-thesis` for the rest), or you have customized their DataCite mappings, you may prefer to keep your current setup or apply only part of this change. Skipping this section does not affect your InvenioRDM installation.

Reviewer: Rewording to map: *what* in 1 paragraph, *why* in another, and *why not* in the last. It also centers the reasoning for the change around accuracy of mapping a thesis to Datacite "Dissertation" as opposed to "Text" given InvenioRDM's adoption of Datacite's definition of Dissertation. (IMHO interoperability is misleading to be frank)  

```suggestion
With InvenioRDM v14, the default resource types "Publication / Thesis" (id `publication-thesis`) and "Publication / Dissertation" (id `publication-dissertation`) were merged into one: "Publication / Thesis" (id `publication-dissertation`). This resource type maps to Datacite's `resourceTypeGeneral` "Dissertation" (`datacite_general: Dissertation` in the InvenioRDM's default resource type YAML file). The separate `publication-thesis` entry is dropped. 

InvenioRDM interprets [Datacite's Dissertation](https://datacite-metadata-schema.readthedocs.io/en/4.7/appendices/appendix-1/resourceTypeGeneral/#dissertation) as covering both former entries, so a single entry mapping to "Dissertation" was more accurate. Datacite's `resourceTypeGeneral` "Text" is not as precise in this context. Staying close to the DataCite schema as a default is a core goal of InvenioRDM.

If you deliberately want to keep both types (for example `publication-dissertation` for PhD work and `publication-thesis` for the rest), or you have customized their DataCite mappings, you may prefer to keep your current setup or apply only part of this change. Skipping this section does not affect your InvenioRDM installation.
```

```bash
curl -LsSf https://raw.githubusercontent.com/inveniosoftware/docs-invenio-rdm/master/docs/releases/v14/migrate_thesis_to_dissertation.py -o /tmp/migrate_thesis_to_dissertation.py
```

```python
# in `invenio shell`
exec(open("/tmp/migrate_thesis_to_dissertation.py").read())
```

Always test against a copy of your data first.

#### Option A: adopt the new default `publication-dissertation`

Use this if you want to switch your `publication-thesis` records over to the new default `publication-dissertation` resource type, id and all. It rewrites your records and drafts, re-registers their DataCite DOIs, and re-indexes them.

1. If you are using a customized list of resource types in `<my_instance>/app_data/vocabularies/resource_types.yaml`, then:
    1. Set `title.<lang>` to "Thesis" (in the appropriate language) for the entry with `id` equal to `publication-dissertation`.

>     2. Remove the entry with `id` equal to `publication-thesis`.

Reviewer: Customizing the list of resource types may imply changing id's (in addition to adding entries e.g.)

```suggestion
    1. Set `title.<lang>` to "Thesis" (in the appropriate language) for the entry with `id` equal or equivalent to `publication-dissertation`.
    2. Remove the entry with `id` equal or equivalent to `publication-thesis`.
```

>     If you didn't customize resource types, you can skip this step.

Reviewer: Given this is in a section that starts with "If you are using a customized list...":

```suggestion
```
